### PR TITLE
chore: ci Initial hive support

### PR DESCRIPTION
**Pull request by bkyryliuk**
### SUMMARY
Hive tests
* added hive to the CI tests via using step and docker compose for sqllab and csv upload tests
* exposed UPLOAD_FOLDER to be able to mount it with the docker
* hdfs for the csv uploads is used instead of s3 via @ pytest.patch

Related cleanup
* reduced presto tests runtime down from 16 to 9 minutes via configuring the poll_interval
* sqlalchemy tables now are dropped as a part of the cleanup
* added ipdb to the test dependencies to be able to debug inside tox environment

Side effects:
* somehow test_update_dataset_update_column started failing for the presto database only on CI, was not able to reproduce it locally, help here would be greatly appreciated 

Followed https://hub.docker.com/r/bde2020/hive/ as an example

### TEST PLAN
Mostly test only changes, CI

### CODE REVIEW

**Comment by eugeniamz** [comment]
:label: database

File: requirements/testing.in [hunk 1]
```diff
@@ -16,7 +16,10 @@
 #
 -r base.in
 -r integration.in
+ipdb
+ipython==7.16.1
```
Comment (john-bodley) [inline requirements/testing.in:20]: Nit. ABC. Also why pin `ipython`?

File: superset/db_engine_specs/hive.py [hunk 2]
```diff
@@ -51,6 +51,27 @@
...
+def upload_to_s3(filename: str, upload_prefix: str, table: Table) -> str:
+    bucket_path = config["CSV_TO_HIVE_UPLOAD_S3_BUCKET"]
+
+    if not bucket_path:
+        logger.info("No upload bucket specified")
+        raise Exception(
+            "No upload bucket specified. You can specify one in the config file."
+        )
+    # Optional dependency
+    import boto3  # pylint: disable=import-error
```
Comment (john-bodley) [inline superset/db_engine_specs/hive.py:63]: Can we place the import at the start of the function?

File: tests/conftest.py [hunk 3]
```diff
@@ -14,18 +14,27 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+# isort:skip_file
```
Comment (john-bodley) [inline tests/conftest.py:17]: Why is this necessary? I prefer the previous logic on 23.

**Comment by villebro** [comment]
FYI @bkyryliuk this needs a rebase

File: .github/workflows/superset-python.yml [hunk 4]
```diff
@@ -152,6 +152,63 @@ jobs:
         run: |
           bash <(curl -s https://codecov.io/bash) -cF python
 
+  test-postgres-hive:
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        # run unit tests in multiple version just for fun
+        python-version: [3.6, 3.7, 3.8]
```
Comment (villebro) [inline .github/workflows/superset-python.yml:160]: I'd suggest dropping 3.6 from new tests going forward.

File: requirements/testing.txt [hunk 5]
```diff
@@ -8,23 +8,39 @@
...
+appnope==0.1.0            # via ipython
 astroid==2.4.2            # via pylint
+backcall==0.2.0           # via ipython
 coverage==5.2.1           # via pytest-cov
+docker==4.3.0             # via -r requirements/testing.in
 flask-testing==0.8.0      # via -r requirements/testing.in
 iniconfig==1.0.1          # via pytest
+ipdb==0.13.3              # via -r requirements/testing.in
+ipython-genutils==0.2.0   # via traitlets
+ipython==7.16.1           # via -r requirements/testing.in, ipdb
```
Comment (villebro) [inline requirements/testing.txt:20]: This is surprising, do we need this dependency?

File: tests/csv_upload_tests.py [hunk 6]
```diff
@@ -241,16 +279,29 @@ def test_import_csv(setup_csv_upload, create_csv_files):
     # make sure that john and empty string are replaced with None
     engine = get_upload_db().get_sqla_engine()
     data = engine.execute(f"SELECT * from {CSV_UPLOAD_TABLE}").fetchall()
-    assert data == [(None, 1, "x"), ("paul", 2, None)]
+    if utils.backend() == "hive":
+        # Be aware that hive only uses first value from the null values list
+        # it can be fixes if we will process csv file on the superset webserver.
+        assert data == [("john", 1, "x"), ("paul", 2, None)]
+    else:
+        assert data == [(None, 1, "x"), ("paul", 2, None)]
```
Comment (villebro) [inline tests/csv_upload_tests.py:287]: I didn't quite understand this, why is "john" showing up here on Hive when other dbs return `None`?

File: tests/db_engine_specs/oracle_tests.py [hunk 7]
```diff
@@ -18,6 +18,7 @@
 from sqlalchemy.dialects import oracle
 from sqlalchemy.dialects.oracle import DATE, NVARCHAR, VARCHAR
 
+from superset.db_engine_specs.hive import HiveEngineSpec
```
Comment (villebro) [inline tests/db_engine_specs/oracle_tests.py:21]: I assume this isn't needed here?

File: tests/model_tests.py [hunk 8]
```diff
@@ -111,44 +111,41 @@ def test_select_star(self):
...
-        LIMIT 100"""
+        elif db.backend == "hive":
+            assert (
+                "SELECT `source`,\n       `target`,\n       `value`\nFROM `energy_usage`\nLIMIT 100"
+                == sql
+            )
+        else:
+            assert (
+                "SELECT source,\n       target,\n       value\nFROM energy_usage\nLIMIT 100"
+                in sql
```
Comment (villebro) [inline tests/model_tests.py:147]: nit/personal preference: I kind of preferred the appearance of `textwrap.dedent` over this.

**Review by villebro — COMMENTED** [review]
I few minor comments, for the most part looks good!

File: requirements/testing.txt (hunk 5 above)
Comment (bkyryliuk) [reply requirements/testing.txt]: this is for ipdb, it CI had issues locationg ipython > 7.16.1, I'll try it again.
As for ipdb - I like using it inside tox, that's why added to the test dependencies. It provides really nice debugging experience.

File: tests/csv_upload_tests.py (hunk 6 above)
Comment (bkyryliuk) [reply tests/csv_upload_tests.py]: see the comment above, hive has it's own csv upload implementation and only supports a single null value where as default implementation works with the list. [ hive engine limitation ]

It can manually process csv to make it feature complete. Updated the comment to be more descriptive & added a todo

File: tests/db_engine_specs/oracle_tests.py (hunk 7 above)
Comment (bkyryliuk) [reply tests/db_engine_specs/oracle_tests.py]: good catch :)

File: .github/workflows/superset-python.yml (hunk 4 above)
Comment (bkyryliuk) [reply .github/workflows/superset-python.yml]: sure thing

File: requirements/testing.txt (hunk 5 above)
Comment (bkyryliuk) [reply requirements/testing.txt:20]: *update* github actions still are unable to install ipython 7.17.0

Comment (villebro) [reply requirements/testing.txt:20]: Nice, I'll try it out 👍

**Review by villebro — APPROVED** [review]
LGTM 👍
